llm/training/config.py

```diff
@@ -0,0 +1,341 @@
+    ):
+        """
+        Event called at the end of training.
+        """
+        self.rsync()
+
+    def on_save(
+        self, args: TrainingArguments, state: TrainerState, control: TrainerControl, **kwargs
+    ):
+        """
+        Event called at the end of training.
+        """
+        self.rsync()
+
+
+def load_config(cls: type, kwargs: Optional[Dict[str, Any]]) -> Any:
+    if kwargs is None:
+        return None
+    return cls(**kwargs)
+
+
+prompt_methods = {}
+prompt_format_methods = {}
+
+
+@dataclass
+class PromptFormatConfig:
+    method: str
+    kwargs: Dict[str, Any] = field(default_factory=dict)
+
+    @classmethod
+    def register(cls, name, method) -> None:
+        prompt_format_methods[name] = method
+
+    def load_role(self, **kwargs):
+        return Role(**kwargs)
+
+    def load(self):
+        method = prompt_format_methods[self.method]
+        kwargs = self.kwargs.copy()
+        for role_field in ['input', 'response', 'system', 'contexts', 'examples']:
+            if role_field in kwargs:
+                kwargs[role_field] = Role(**kwargs[role_field])
+        return method(**kwargs)
+
+
+PromptFormatConfig.register(PromptFormat.__name__, PromptFormat)
+PromptFormatConfig.register(UserAssistantFormat.__name__, UserAssistantFormat)
+PromptFormatConfig.register(RedpajamaFormat.__name__, RedpajamaFormat)
+PromptFormatConfig.register(VicunaFormat.__name__, VicunaFormat)
+PromptFormatConfig.register(Llama2Format.__name__, Llama2Format)
+PromptFormatConfig.register(TogetherLlama2Format.__name__, TogetherLlama2Format)
+PromptFormatConfig.register(ChatMLFormat.__name__, ChatMLFormat)
+PromptFormatConfig.register(DollyFormat.__name__, DollyFormat)
+
+@dataclass
+class PromptConfig:
+    method: str
+    kwargs: Dict[str, Any] = field(default_factory=dict)
+
+    @classmethod
+    def register(cls, name, method) -> None:
+        prompt_methods[name] = method
+
+    def load(self, format):
+        method = prompt_methods[self.method]
+        kwargs = self.kwargs.copy()
+        if "examples" in kwargs:
+            kwargs["examples"] = [Message(**x) for x in kwargs["examples"]]
+        return method(format=format, **kwargs)
+
+
+PromptConfig.register(Prompt.__name__, Prompt)
+PromptConfig.register(ZeroShotQA.__name__, ZeroShotQA)
+PromptConfig.register(FewShotQA.__name__, FewShotQA)
+PromptConfig.register(StandaloneQuestion.__name__, StandaloneQuestion)
+
+
+dataset_writers = {}
+
+
+@dataclass
+class DatasetWriterConfig:
+    method: str
+    kwargs: Dict[str, Any] = field(default_factory=dict)
+
+    @classmethod
+    def register(cls, name, method):
+        dataset_writers[name] = method
+
+    def save(self, dataset: Dataset):
+        method = dataset_writers[self.method]
+        method(dataset, **self.kwargs)
+
+
+DatasetWriterConfig.register(
+    "save_to_disk", lambda dataset, dataset_path: dataset.save_to_disk(dataset_path)
+)
```

Comment:
might be nice to have some of the basic `to_json` `to_csv` options registered here

Reply:
yep, will add a bunch later today